### PR TITLE
Updating IMDS endpoint discovery to be non-throwing

### LINKED ISSUE #8790: Consider MSI endpoint detection without sending an HTTP request
By doing something like:

```
try
    {
        using(var client = new TcpClient())
        {
            var result = client.BeginConnect(host, port, null, null);
            var success = result.AsyncWaitHandle.WaitOne(timeout);
            client.EndConnect(result);
            return success;
        }
    }
    catch
    {
        return false;
    }
```

It would `OperationCancelledException` on every startup.

### LINKED ISSUE #9550: [BUG] 500ms timeout on IMDS available probe is too aggressive
**Describe the bug**
When using `ManagedIdentityCredential` to run in the context of a Managed Identity in Azure, a probe is made to `http://169.254.169.254/metadata/identity/oauth2/token` to test whether a managed identity is available or not.  The timeout on this request is 500ms and the response (or lack of) is cached and cannot be cleared/refreshed.  The 500ms timeout cannot be configured.
https://github.com/Azure/azure-sdk-for-net/blob/b0ff97d220ec1e83d43767d693b17a3b6245314e/sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs#L22

On a weak or busy instance, this causes the Msi to be marked as unavailable and it cannot be cleared without relaunching the process.

***Exception or Stack Trace***
https://github.com/Azure/azure-sdk-for-net/blob/b0ff97d220ec1e83d43767d693b17a3b6245314e/sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs#L122

**To Reproduce**
Attempt to use managed identities on a stressed server, or use a proxy to intercept the probe to ensure the response is not received in 500ms.

***Code Snippet***
```
                var accountUri = new Uri("https://storageaccountname.blob.core.windows.net/");
                var client = new BlobServiceClient(accountUri, new ManagedIdentityCredential());
                var containerClient = client.GetBlobContainerClient("container");
                var blobClient = containerClient.GetBlobClient("blob");

                await blobClient.DownloadToAsync(new MemoryStream());
```

**Expected behavior**
* `blobClient` should use the Managed Identity that is available to it.
* It should be possible to configure the 500ms timeout.
* It should be possible to refresh/reset the Msi status in the case of an error without having to restart the process

**Setup (please complete the following information):**
 - Docker Image: mcr.microsoft.com/windows/servercore:ltsc2019
 - Azure.Identity: 1.1.0

### LINKED ISSUE #9588: [BUG] DefaultAzureCredential fails in production on Azure VM
**Describe the bug**
`DefaultAzureCredential` fails to find the managed identity endpoint in a production build on an Azure VM (there is a rare chance it succeeds). The unchanged code does not fail when debugging in Visual Studio on the exact same VM.

On my dev machine, `DefaultAzureCredential` will successfully use an `EnvironmentCredential` instead of `ManagedIdentityCredential`. As a temporary workaround, I replicated the same environment variables on the target VM, but `DefaultAzureCredential` *could not find those environment variables either* (I set them as system variables instead of user variables on the target VM to ensure Azure's Compute extensions for remote powershell scripts will have access to them).

I happen to know that MSI is functioning properly on the target VM because AzCopy manages to successfully authenticate using MSI earlier in the exact same script.

***Exception or Stack Trace***
Logs as recorded by `AzureEventSourceListener`
```
Azure.Identity.DefaultAcureCredential.GetToken invoked. Scopes: [ https://storage.azure.com/.default ] ParentRequestId: 432149f7-cbac-4c24-880e-ac328b406736
Azure.Identity.EnvironmentCredential.GetToken invoked. Scopes: [ https://storage.azure.com/.default ] ParentRequestId: 432149f7-cbac-4c24-880e-ac328b406736
Azure.Identity.EnvironmentCredential.GetToken was unable to retrieve an access token. Scopes: [ https://storage.azure.com/.default ] ParentRequestId: 432149f7-cbac-4c24-880e-ac328b406736 Exception: Azure.Identity.CredentialUnavailableException (0x80131500): Environment variables not fully configured. AZURE_TENANT_ID and AZURE_CLIENT_ID must be set, along with either AZURE_CLIENT_SECRET or AZURE_USERNAME and AZURE_PASSWORD. Currently set variables [  ]
Azure.Identity.ManagedIdentityCredential.GetToken invoked. Scopes: [ https://storage.azure.com/.default ] ParentRequestId: 432149f7-cbac-4c24-880e-ac328b406736
Request [ec635421-038f-474a-98b5-e17014f36494] GET http://169.254.169.254/metadata/identity/oauth2/token??pi-version=REDACTED 
x-ms-client-request-id:ec635421-038f-474a-98b5-e17014f36494
x-ms-return-client-request-id:true
User-Agent:azsdk-net-Identity/1.0.0+809e5374b6b040adcb6425f9e46b19f5182b3488,(.NET Core 4.6.28008.02; Microsoft Windows 10.0.17763 )

Azure.Identity.ManagedIdentityCredential.GetToken was unable to retrieve an access token. Scopes: [ https://storage.azure.com/.default ] ParentRequestId: 432149f7-cbac-4c24-880e-ac328b406736 Exception: Azure.Identity.CredentialUnavailableException (0x80131500): No managed identity endpoint found.
Azure.Identity.SharedTokenCacheCredential.GetToken invoked. Scopes: [ https://storage.azure.com/.default ] ParentRequestId: 432149f7-cbac-4c24-880e-ac328b406736
Azure.Identity.SharedTokenCacheCredential.GetToken was unable to retrieve an access token. Scopes: [ https://storage.azure.com/.default ] ParentRequestId: 432149f7-cbac-4c24-880e-ac328b406736 Exception: Azure.Identity.CredentialUnavailableException (0x80131500): No accounts were discovered in the shared token cache. To fix, authenticate through tooling supporting azure developer sign on.
Azure.Identity.DefaultAcureCredential.GetToken was unable to retrieve an access token. Scopes: [ https://storage.azure.com/.default ] ParentRequestId: 432149f7-cbac-4c24-880e-ac328b406736 Exception: Azure.Identity.AuthenticationFailedException (0x80131500): The DefaultAzureCredential failed to retrieve a token from the included credentials.
  EnvironmentCredential is unavailable Environment variables not fully configured. AZURE_TENANT_ID and AZURE_CLIENT_ID must be set, along with either AZURE_CLIENT_SECRET or AZURE_USERNAME and AZURE_PASSWORD. Currently set variables [  ].
  ManagedIdentityCredential is unavailable No managed identity endpoint found..
  SharedTokenCacheCredential is unavailable No accounts were discovered in the shared token cache. To fix, authenticate through tooling supporting azure developer sign on..
See inner exception for more detail.
 ---> System.AggregateException (0x80131500): The DefaultAzureCredential failed to retrieve a token from the included credentials. (Environment variables not fully configured. AZURE_TENANT_ID and AZURE_CLIENT_ID must be set, along with either AZURE_CLIENT_SECRET or AZURE_USERNAME and AZURE_PASSWORD. Currently set variables [  ]) (No managed identity endpoint found.) (No accounts were discovered in the shared token cache. To fix, authenticate through tooling supporting azure developer sign on.)
 ---> Azure.Identity.CredentialUnavailableException (0x80131500): Environment variables not fully configured. AZURE_TENANT_ID and AZURE_CLIENT_ID must be set, along with either AZURE_CLIENT_SECRET or AZURE_USERNAME and AZURE_PASSWORD. Currently set variables [  ]
```

**To Reproduce**
I do not have a succinct, guaranteed reproduction at this time. This error occurs in release builds run on Azure VMs, and these builds are being invoked through Azure Compute's extensions to run powershell commands on Azure VMs.

***Code Snippet***
```csharp
var serviceClient = new BlobServiceClient(
    myStorageEndpoint,
    new Azure.Identity.DefaultAzureCredential());
var containerClient = serviceClient.GetBlobContainerClient(myNewContainerName);
containerClient.Create(); //throws
```

**Expected behavior**
DefaultAzureCredential should successfully obtain a token for my requests.

**Screenshots**
N/A

**Setup:**
 - OS: Windows Datacenter 2019
 - IDE : Release build only (occasionally crops up in Visual Studio 2017)
 - Version of the Library used: Identity 1.0.0; Storage 12.0.0

**Additional context**
N/A

**Information Checklist**
Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Bug Description Added
- [x] Repro Steps Added
- [x] Setup information Added

### PR DESCRIPTION
**Pull request by schaabs**
This updates the mechanism to discover if the IMDS endpoint is available from trying to send a request to directly trying to create a TCP connection. This enables us to check for existence without throwing and handling exceptions.

*  Update Managed Identity discovery to be instance instance based
*  Updating IMDS endpoint discovery to be non-throwing
*  Increase IMDS discovery from 500ms to 1000ms
*  Updating ManagedIdentity credential to raise CredentialUnavailableException on IMDS 400 response
*  Adding recordable tests for ManagedIdentityCredenital IMDS authentication

Fixes #8790 
Fixes #9550 
Fixes #9588

### CODE REVIEW

**Comment by pakrym** [comment]
It's unfortunate that we don't have positive tests for these.

File: sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs [hunk 1]
```diff
@@ -206,86 +209,48 @@ private Request CreateAuthRequest(MsiType msiType, string[] scopes, string clien
...
-                try
+                using (var client = new TcpClient())
                 {
-                    AzureIdentityEventSource.Singleton.ProbeImdsEndpoint(request.Uri);
-
-                    Response response = await _pipeline.HttpPipeline.SendRequestAsync(request, CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, imdsTimeout).Token).ConfigureAwait(false);
+                    var result = client.BeginConnect(s_imdsHostIp, s_imdsPort, null, null);
 
-                    AzureIdentityEventSource.Singleton.ImdsEndpointFound(request.Uri);
+                    var success = await Task.Run<bool>(() => result.AsyncWaitHandle.WaitOne(ImdsAvailableTimeoutMs), cancellationToken).ConfigureAwait(false);
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs]: There is a more efficient way of doing this https://docs.microsoft.com/en-us/dotnet/standard/asynchronous-programming-patterns/interop-with-other-asynchronous-patterns-and-types?redirectedfrom=MSDN#WHToTap but I'm not sure how much we care.

**Comment by pakrym** [comment]
Are you planning to add the HTTP request back and use TcpConnection as optimistic check? I was postponing the full review waiting for the change.

**Comment by schaabs** [comment]
@pakrym After thinking about it further I'm unconvinced that the HTTP request adds value over just the Tcp connection. Since we're only checking for a response not a sucessful response, the only corner case that this would be that something else is listening on this private ip at the correct port that isn't IMDS and doesn't use HTTP. So I don't think the added complexity is worth guarding this very edge case. Plus there is a workaround in this case for the `DefaultAzureCredential` as the customer could disable the `ManagedIdentityCredential`.

File: sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs [hunk 2]
```diff
@@ -206,86 +209,48 @@ private Request CreateAuthRequest(MsiType msiType, string[] scopes, string clien
...
-
-                request.Uri.AppendQuery("api-version", ImdsApiVersion);
-
-                CancellationToken imdsTimeout = new CancellationTokenSource(ImdsAvailableTimeoutMs).Token;
-
-                try
+                using (var client = new TcpClient())
                 {
-                    Response response = _pipeline.HttpPipeline.SendRequest(request, CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, imdsTimeout).Token);
+                    var result = client.BeginConnect(s_imdsHostIp, s_imdsPort, null, null);
```
Comment (KrzysztofCwalina) [inline sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs]: I think it would be better to use Connect.

File: sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs [hunk 3]
```diff
@@ -206,86 +209,48 @@ private Request CreateAuthRequest(MsiType msiType, string[] scopes, string clien
...
-
-                CancellationToken imdsTimeout = new CancellationTokenSource(ImdsAvailableTimeoutMs).Token;
-
-                try
+                using (var client = new TcpClient())
                 {
-                    AzureIdentityEventSource.Singleton.ProbeImdsEndpoint(request.Uri);
-
-                    Response response = await _pipeline.HttpPipeline.SendRequestAsync(request, CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, imdsTimeout).Token).ConfigureAwait(false);
+                    var result = client.BeginConnect(s_imdsHostIp, s_imdsPort, null, null);
```
Comment (KrzysztofCwalina) [inline sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs]: I think it would be better to use ConnectAsync

File: sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs (hunk 1 above)
Comment (KrzysztofCwalina) [reply sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs]: Do we care about the optimization? The more efficient way complicates the code a bit.

**Review by KrzysztofCwalina — COMMENTED** [review]
I think it's good to avoid the exception being logged. The exception is just an implementation detail irrelevant to the user of the SDK. But, we can use Connect/ConnectAsync directly to avoid the exception being logged, but still use modern APIs. I worry about using the old async pattern in our APIs as some of these calls are implemented in a bit of a hacky way in .NET Core.

File: sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs (hunk 1 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs]: I agree. Since this should only be called once per process I think that we can leave the simplified code here and optimize later on if we find it to be a problem.

File: sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs (hunk 2 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs]: After looking into this we can't use `Connect` and `ConnectAsync` because they don't provide a timeout or CancellationToken parameter.

File: sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs (hunk 3 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs]: After looking into this we can't use `Connect` and `ConnectAsync` because they don't provide a timeout or CancellationToken parameter.

File: sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs [hunk 4]
```diff
@@ -206,86 +209,48 @@ private Request CreateAuthRequest(MsiType msiType, string[] scopes, string clien
...
 
-                    return true;
-                }
-                // we only want to handle the case when the imdsTimeout resulted in the request being canceled.
-                // this indicates that the request timed out and that imds is not available.  If the operation
-                // was user canceled we don't wan't to handle the exception so s_identityAvailable will
-                // remain unset, as we still haven't determined if the imds endpoint is available.
-                catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
-                {
-                    AzureIdentityEventSource.Singleton.ImdsEndpointUnavailable(request.Uri);
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs]: We are not calling this event anywhere. Should it be removed(I don't think it shoudl)? Same for one above

File: sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialImdsLiveTests.cs [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    {
+        public ManagedIdentityCredentialImdsLiveTests(bool isAsync) : base(isAsync)
+        {
+            Sanitizer = new IdentityRecordedTestSanitizer();
+        }
+
+        [SetUp]
+        public void ResetManagedIdenityClient()
+        {
+            typeof(ManagedIdentityClient).GetField("s_msiType", BindingFlags.NonPublic | BindingFlags.Static).SetValue(null, 0);
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialImdsLiveTests.cs]: Can we add an internal method for this? This piece of reflection gets copied around a lot.

File: sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs [hunk 6]
```diff
@@ -18,8 +20,10 @@ internal class ManagedIdentityClient
 
         // IMDS constants. Docs for IMDS are available here https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http
         private static readonly Uri s_imdsEndpoint = new Uri("http://169.254.169.254/metadata/identity/oauth2/token");
+        private static readonly IPAddress s_imdsHostIp = IPAddress.Parse("169.254.169.254");
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs]: Should this take into account `MSI_ENDPOINT` variable?

File: sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs (hunk 4 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs]: No, good catch I'll add this call back in.

File: sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs (hunk 6 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs]: No this endpoint is specific to IMDS,  not other managed identity environments which use MSI_ENDPOINT.

File: sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialImdsLiveTests.cs (hunk 5 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialImdsLiveTests.cs]: The need for this has been removed since I've made the the discovery instance based.

File: sdk/identity/Azure.Identity/src/CredentialPipeline.cs [hunk 7]
```diff
@@ -69,5 +70,13 @@ public CredentialPipeline WithAuthorityHost(Uri authorityHost)
         {
             return new CredentialPipeline(authorityHost, HttpPipeline, Diagnostics);
         }
+
+        private class CredentialResponseClassifier : ResponseClassifier
+        {
+            public override bool IsRetriableResponse(HttpMessage message)
+            {
+                return base.IsRetriableResponse(message) || message.Response.Status == 404;
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/src/CredentialPipeline.cs]: Interesting, would retrying multiple times in a row help when server returns 404?

Comment (schaabs) [reply sdk/identity/Azure.Identity/src/CredentialPipeline.cs]: Yes apparently the service returns 404 when it is upgrading. https://docs.microsoft.com/en-us/azure/active-directory/managed-identities-azure-resources/how-to-use-vm-token#error-handling

File: sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs [hunk 8]
```diff
@@ -15,74 +18,121 @@ internal class ManagedIdentityClient
...
-        public virtual AccessToken Authenticate(MsiType msiType, string[] scopes, string clientId, CancellationToken cancellationToken)
+        protected string ClientId { get; }
+
+        public virtual ExtendedAccessToken Authenticate(string[] scopes, CancellationToken cancellationToken)
         {
-            using Request request = CreateAuthRequest(msiType, scopes, clientId);
+            MsiType msiType = GetMsiType(cancellationToken);
+
+            // if msi is unavailable or we were unable to determine the type return CredentialUnavailable exception that no endpoint was found
+            if (msiType == MsiType.Unavailable || msiType == MsiType.Unknown)
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs]: Why would we return Unknown from GetMsiType?

File: sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs [hunk 9]
```diff
@@ -102,49 +152,51 @@ public virtual MsiType GetMsiType(CancellationToken cancellationToken)
...
+                            _msiType = MsiType.Imds;
                         }
-                        // if MSI_ENDPOINT is NOT set and IMDS enpoint is not available ManagedIdentity is not available
+                        // if MSI_ENDPOINT is NOT set and IMDS endpoint is not available ManagedIdentity is not available
                         else
                         {
-                            s_msiType = MsiType.Unavailable;
+                            _msiType = MsiType.Unavailable;
+
+                            AzureIdentityEventSource.Singleton.ImdsEndpointUnavailable(s_imdsEndpoint);
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs]: I think we'll end up calling ImdsEndpointUnavailable twice, once in `ImdsAvailable` and second time here.

File: sdk/identity/Azure.Identity/src/CredentialPipeline.cs (hunk 7 above)
Comment (pakrym) [reply sdk/identity/Azure.Identity/src/CredentialPipeline.cs]: 🤦‍♂

File: sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs (hunk 9 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs]: Thanks I meant to remove this.

File: sdk/identity/Azure.Identity/tests/Azure.Identity.Tests.csproj [hunk 10]
```diff
@@ -11,6 +11,7 @@
     <PackageReference Include="Microsoft.NET.Test.Sdk" />
     <PackageReference Include="Moq" />
     <PackageReference Include="BenchmarkDotNet" />
+    <PackageReference Include="Azure.Security.KeyVault.Secrets" />
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/tests/Azure.Identity.Tests.csproj]: Why is this required?

File: sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialLiveTests.cs [hunk 11]
```diff
@@ -6,6 +6,8 @@
 using System.Threading.Tasks;
 using System.Reflection;
 using Azure.Core;
+using System;
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialLiveTests.cs]: rem

File: sdk/identity/Azure.Identity/tests/Azure.Identity.Tests.csproj (hunk 10 above)
Comment (pakrym) [reply sdk/identity/Azure.Identity/tests/Azure.Identity.Tests.csproj]: Found it

File: sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs (hunk 8 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/src/ManagedIdentityClient.cs]: Unknown should not be returned. However, since `GetMsiType` returns the member variable `_msiType` I have the check for `MsiType.Unknown` I have this check to guard against any possible races with other setters of this field.

File: sdk/identity/Azure.Identity/tests/Azure.Identity.Tests.csproj (hunk 10 above)
Comment (schaabs) [reply sdk/identity/Azure.Identity/tests/Azure.Identity.Tests.csproj]: it's required for the tests because we use Key Vault to validate that the tokens we get are valid.

File: sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialTests.cs [hunk 12]
```diff
@@ -79,17 +74,19 @@ public async Task VerifyImdsRequestWithClientIdMockAsync()
...
+                var pipeline = CredentialPipeline.GetInstance(options);
+
+                ManagedIdentityCredential credential = InstrumentClient(new ManagedIdentityCredential(pipeline, new MockManagedIdentityClient(pipeline, "mock-client-id") { ImdsAvailableFunc = _ => true }));
 
                 AccessToken actualToken = await credential.GetTokenAsync(new TokenRequestContext(MockScopes.Default));
 
                 Assert.AreEqual(expectedToken, actualToken.Token);
 
-                MockRequest request = mockTransport.Requests[mockTransport.Requests.Count - 1];
+                MockRequest request = mockTransport.Requests[0];
```
Comment (pakrym) [inline sdk/identity/Azure.Identity/tests/ManagedIdentityCredentialTests.cs]: You can use SingleRequest property here

**Review by pakrym — APPROVED** [review]

**Comment by schaabs** [comment]
Closing and reopening to reset the CI.
